### PR TITLE
Benchmarked Geoid postprocessor

### PR DESCRIPTION
**Pull request by Shangxin-Liu**
Here is the newest geoid code with detailed comments. @gassmoeller @bangerth @ian-r-rose 

It was already benchmarked against the analytical solution of both single degree order and multiple degree order input density field. The numerical solution of the 3D density integral which working wth AMR is the same as analytic solution (even in coarse mesh like global refinement 0 the two solutions are pretty close because the integral is based on the volume Guass quadrature points). The spherical harmonics of dynamic topography were tested against the grid output from dynamic topography postprocessor, i.e., the sum up of all degree order components equal the grid dynamic topography results.

I add a test prm file using 1 global refinement degree here to input only degree 2 order 0 density perturbation from a constant density (i.e., degree 0 order 0 component), the output density contribution spherical harmonics have the same values as the analytical solutions. It's a pure mathematic test of the single degree order density input with no physical meaning so most of the material property parameter is set to 1 to make this easily achieved.

The geoid code allows the user to input the min and max degree, whether to output results in cartesian coordinates or lon&lat coordinates, density above, density below, and whether also output the spherical harmonic coefficients of total geoid, density contribution, surface topography, and CMB topography.

### CODE REVIEW

**Comment by gassmoeller** [comment]
I did not have the time to review this PR this evening, but it is number one on my list for today.

File: source/postprocess/geoid.cc [hunk 1]
```diff
@@ -0,0 +1,1025 @@
...
+      std::vector<double> CMB_dyna_topo_contribution_coecos; // a vector to store cos terms of CMB dynamic topography contribution SH coefficients
+      std::vector<double> CMB_dyna_topo_contribution_coesin; // a vector to store sin terms of CMB dynamic topography contribution SH coefficients
+      std::vector<double> geoid_coecos;  // a vector to store cos terms of geoid anomaly SH coefficients
+      std::vector<double> geoid_coesin;  // a vector to store sin terms of geoid anomaly SH coefficients
+
+      // First compute the spherical harmonic contributions from density anomaly, surface dynamic topography and CMB dynamic topography
+      int ind = 0; // coefficients index
+      for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+        {
+          for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: use `unsigned int` for the counter variables. Maybe you will need to convert it to a different type to use it in the formulas below.

File: source/postprocess/geoid.cc [hunk 2]
```diff
@@ -0,0 +1,1025 @@
...
+              surface_dyna_topo_contribution_coesin.push_back(coesin_surface_dyna_topo);
+
+              double coecos_CMB_dyna_topo = (4 * numbers::PI * G / (surface_gravity * (2 * ideg + 1)))
+                                            * CMB_delta_rho*SH_CMB_dyna_topo_coes.second.first.at(ind)*inner_radius*std::pow(inner_radius/outer_radius,ideg+1);
+              double coesin_CMB_dyna_topo = (4 * numbers::PI * G / (surface_gravity * (2 * ideg + 1)))
+                                            * CMB_delta_rho*SH_CMB_dyna_topo_coes.second.second.at(ind)*inner_radius*std::pow(inner_radius/outer_radius,ideg+1);
+              CMB_dyna_topo_contribution_coecos.push_back(coecos_CMB_dyna_topo);
+              CMB_dyna_topo_contribution_coesin.push_back(coesin_CMB_dyna_topo);
+
+              ind +=1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: this is simpler if you type: `++ind;`

File: source/postprocess/geoid.cc [hunk 3]
```diff
@@ -0,0 +1,1025 @@
...
+            }
+        }
+
+      // Then sum the three contributions together to get the spherical harmonic coefficients of geoid anomaly
+      ind = 0; // coefficients index
+      for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+        {
+          for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `unsigned int` in both loops

File: source/postprocess/geoid.cc [hunk 4]
```diff
@@ -0,0 +1,1025 @@
...
+        {
+          const std_cxx11::array<double,dim> scoord = aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(surface_cell_locations.at(i));
+          const double phi = scoord[1];
+          const double theta = scoord[2];
+          surface_cell_spherical_coordinates.push_back(std::make_pair(theta,phi));
+        }
+
+      // Compute the geoid anomaly based on spherical harmonics
+      std::vector<double> geoid_anomaly;
+      double prefact;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: same as above, remove `prefact` here

File: source/postprocess/geoid.cc [hunk 5]
```diff
@@ -0,0 +1,1025 @@
...
+          int ind = 0;
+          double geoid_value = 0;
+          for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+            {
+              for (int iord = 0; iord < ideg+1; iord++)
+                {
+                  const double cos_component = boost::math::spherical_harmonic_r(ideg,iord,surface_cell_spherical_coordinates.at(i).first,surface_cell_spherical_coordinates.at(i).second); //real / cos part
+                  const double sin_component = boost::math::spherical_harmonic_i(ideg,iord,surface_cell_spherical_coordinates.at(i).first,surface_cell_spherical_coordinates.at(i).second); //imaginary / sine part
+                  // normalization after Dahlen and Tromp, 1986, Appendix B.6
+                  if (iord == 0)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: replace these lines by the same suggestion as above

File: source/postprocess/geoid.cc [hunk 6]
```diff
@@ -0,0 +1,1025 @@
...
+                    {
+                      prefact = 1.;
+                    }
+                  else
+                    {
+                      prefact = sqrt(2.);
+                    }
+                  geoid_value += prefact*(geoid_coecos.at(ind)*cos_component+geoid_coesin.at(ind)*sin_component);
+                  ind += 1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `++ind;`

File: source/postprocess/geoid.cc [hunk 7]
```diff
@@ -0,0 +1,1025 @@
...
+        {
+          // have a stream into which we write the SH coefficients data from density anomaly contribution.
+          // The text stream is then later sent to processor 0
+          std::ostringstream output_density_anomaly_contribution_SH_coes;
+
+          // Prepare the output SH coefficients data from density anomaly contribution.
+          unsigned int SH_coes_ind = 0;
+          for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+            {
+              for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: use `unsigned int`

File: source/postprocess/geoid.cc [hunk 8]
```diff
@@ -0,0 +1,1025 @@
...
+                {
+                  output_density_anomaly_contribution_SH_coes << ideg
+                                                              << ' '
+                                                              << iord
+                                                              << ' '
+                                                              << density_anomaly_contribution_coecos.at(SH_coes_ind)
+                                                              << ' '
+                                                              << density_anomaly_contribution_coesin.at(SH_coes_ind)
+                                                              << std::endl;
+                  SH_coes_ind += 1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: use `++`

File: source/postprocess/geoid.cc [hunk 9]
```diff
@@ -0,0 +1,1025 @@
...
+        {
+          // have a stream into which we write the SH coefficients data from surface dynamic topography contribution.
+          // The text stream is then later sent to processor 0
+          std::ostringstream output_surface_dynamic_topo_contribution_SH_coes;
+
+          // Prepare the output SH coefficients data from surface dynamic topography contribution.
+          unsigned int SH_coes_ind = 0;
+          for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+            {
+              for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `unsigned int`

File: source/postprocess/geoid.cc [hunk 10]
```diff
@@ -0,0 +1,1025 @@
...
+                {
+                  output_surface_dynamic_topo_contribution_SH_coes << ideg
+                                                                   << ' '
+                                                                   << iord
+                                                                   << ' '
+                                                                   << surface_dyna_topo_contribution_coecos.at(SH_coes_ind)
+                                                                   << ' '
+                                                                   << surface_dyna_topo_contribution_coesin.at(SH_coes_ind)
+                                                                   << std::endl;
+                  SH_coes_ind += 1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `++ind`

File: source/postprocess/geoid.cc [hunk 11]
```diff
@@ -0,0 +1,1025 @@
...
+        {
+          // have a stream into which we write the SH coefficients data from CMB dynamic topography contribution.
+          // The text stream is then later sent to processor 0
+          std::ostringstream output_CMB_dynamic_topo_contribution_SH_coes;
+
+          // Prepare the output SH coefficients data from CMB dynamic topography contribution.
+          unsigned int SH_coes_ind = 0;
+          for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+            {
+              for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `unsigned int`

File: source/postprocess/geoid.cc [hunk 12]
```diff
@@ -0,0 +1,1025 @@
...
+                {
+                  output_CMB_dynamic_topo_contribution_SH_coes << ideg
+                                                               << ' '
+                                                               << iord
+                                                               << ' '
+                                                               << CMB_dyna_topo_contribution_coecos.at(SH_coes_ind)
+                                                               << ' '
+                                                               << CMB_dyna_topo_contribution_coesin.at(SH_coes_ind)
+                                                               << std::endl;
+                  SH_coes_ind += 1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `++ind`

File: source/postprocess/geoid.cc [hunk 13]
```diff
@@ -0,0 +1,1025 @@
...
+        {
+          // have a stream into which we write the geoid anomaly SH coefficients data.
+          // The text stream is then later sent to processor 0
+          std::ostringstream output_geoid_anomaly_SH_coes;
+
+          // Prepare the output geoid anomaly SH coefficients data
+          unsigned int SH_coes_ind = 0;
+          for (int ideg =  min_degree; ideg < max_degree+1; ideg++)
+            {
+              for (int iord = 0; iord < ideg+1; iord++)
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `unsigned int`

File: source/postprocess/geoid.cc [hunk 14]
```diff
@@ -0,0 +1,1025 @@
...
+                {
+                  output_geoid_anomaly_SH_coes << ideg
+                                               << ' '
+                                               << iord
+                                               << ' '
+                                               << geoid_coecos.at(SH_coes_ind)
+                                               << ' '
+                                               << geoid_coesin.at(SH_coes_ind)
+                                               << std::endl;
+                  SH_coes_ind += 1;
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: `++ind`

File: source/postprocess/geoid.cc [hunk 15]
```diff
@@ -0,0 +1,1025 @@
...
+                            Patterns::Integer (0),
+                            "This parameter normally is set to 2 since the perturbed gravitational potential at degree 1 always vanishes"
+                            "in a reference frame with the planetary center of mass as the center of the reference frame.");
+          prm.declare_entry("Output data in geographical coordinates", "false",
+                            Patterns::Bool(),
+                            "Option to output the geoid anomaly in geographical coordinates (latitude and longitude). "
+                            "The default is false, so postprocess will output the data in geocentric coordinates (x,y,z) as normally.");
+          prm.declare_entry("Density above","0",
+                            Patterns::Double (0),
+                            "The density value out of the surface boundary.");
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: I would call this `The density value above the surface boundary`

File: source/postprocess/geoid.cc [hunk 16]
```diff
@@ -0,0 +1,1025 @@
...
+          prm.declare_entry("Density below","8000",
+                            Patterns::Double (0),
+                            "The density value out of the CMB boundary.");
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: Same here `below the CMB boundary`

File: source/postprocess/geoid.cc [hunk 17]
```diff
@@ -0,0 +1,1025 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    ASPECT_REGISTER_POSTPROCESSOR(Geoid,
+                                  "geoid",
+                                  "A postprocessor that computes a measure of geoid anomaly based on the density anomaly determined "
+                                  "from the temperature field in the mantle, and the dynamic topography at the surface and core mantle "
+                                  "boundary(CMB). The geoid is computed form the spherical harmonics expansion, so the geometry of the "
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: "computed *from* ..."

File: tests/geoid.prm [hunk 18]
```diff
@@ -0,0 +1,113 @@
+# A simple pure math setup to compute geoid in single degree 2 order 0 perturbation density field.
+# The perturbation is added to a constant density, i.e., degree 0 order 0 component.
+# The analytic solution of the denbsity integral contribution of this case should be 26.529660019
```
Comment (gassmoeller) [inline tests/geoid.prm]: density

File: tests/geoid.prm [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+# for (2,0) component and 79.8723303437 for degree 0 order 0, and the sum up of the spherical harmonics
```
Comment (gassmoeller) [inline tests/geoid.prm]: .. the sum of ...

**Comment by Shangxin-Liu** [comment]
This is the final version after Rene's comments. @gassmoeller  A new change file has been added into modules/changes including Ian @ian-r-rose  also as a contributor. The gnuplot visualization file is removed. I also change to use the spherical harmonics in utilities.cc instead of from boost so we don't need any normalization prefactor since it's already done in utilities.cc.

**Comment by ian-r-rose** [comment]
Have you benchmarked the computation against a total geoid computation (i.e., Zhong et al, 2008?) I found that one of the most challenging things was combining all the contributions to the geoid in the correct proportions.

**Comment by Shangxin-Liu** [comment]
This benchmark is a global input degree 2 order 0 (I also have multiple degree order input density field benchmarked but for the test prm file, let's use a single degree order for now), The density integral numerical solution is very close to the analytic solution, even in coarse mesh. The calculation method of dynamic topography contribution is the same as dynamic topography postprocessor and the results are further transferred to spherical harmonic domain. The test of dynamic topography contribution was done in the sense that the sum of the spherical harmonics from dynamic topography is the same as the grid output from dynamic topography postprocessor. So now the accuracy of dynamic topography and geoid is essentially the same thing. When we find the best accuracy algorithm (boundary volume density/stress/pressure, or boundary face density/pressure and volume stress, or consistent boundary flux) in dynamic topography, we can also use the same algorithm to calculate dynamic topography contribution in geoid, but we should always keep the dynamic topography calculation method the same in geoid and dynamic topography postprocessor. 

I already saw several future improvements of geoid but may not finish in the following days:
1. Directly call dynamic topography postprocessor to calculate the topography contribution. I think it's better to make this change after we find the best accuracy algorithm to calculate dynamic topography, and then make a significant change to the code interface since the dynamic topography contribution needs other extra calculations from dynamic topography.
2. Move the spherical harmonic expansion function into utilities.cc. This may need us to discuss the universe format of the input spherical function, another pull request.

I've not benchmarked against Zhong 2008. It's my current struggle:-)  Maybe we can also let this benchmark results to decide which dynamic topography calculation method is the best. @ian-r-rose  @gassmoeller

**Comment by gassmoeller** [comment]
I agree with Shangxin. As far as I understood the postprocessor passes a very simple analytical solution. I would not make the merge dependent on the Zhong 2008 results, especially given how much Scott and Grant (and previously Timo and I) struggled to reproduce those results (see also Scott's mail on the mailing list). I will just push the button for now, and I am happy about any follow-up improvement.

**Comment by ian-r-rose** [comment]
Sounds good, congratulations @Shangxin-Liu! I have also had a heck of a time reproducing the Zhong et al (2008) results. To your specific points:

1. I am still working on the CBF topography benchmarking. In 2D it seems to be significantly more accurate. Once that is working we can have the geoid postprocessor directly get the dynamic topography from the dynamic topography plugin (to avoid redundant calculations), as is done in the visualization postprocessor in #1572 .

2. This makes good sense to me.

@gassmoeller I think that the multipole approach of #1060 for the is still a good idea, but if I pick it back up I will incorporate it into the current geoid postprocessor.

**Comment by Shangxin-Liu** [comment]
Yes. I'm a bit conservative:-) Let us prove that in ASEPCT's 3D problem the CBF approach is indeed better than the previous one and then make this significant change. @gassmoeller 

Ian @ian-r-rose , I've showed my notebook to you how I transferred the layer integral of density spherical harmonics to 3D global integral directly to work around AMR. Is it the same as your multipole approach? I may not be familiar with the terminology but the math nature should be the same, yes? Another possible follow-up change may be to also make geoid work in 2D problem although now very few people interested in 2D geoid and the geoid from a regional cap since it's only part of the total geoid and it's hard to compare that to large scale geodesy gravity model. We may put Ian's 2D geoid stuff into the code but let's see how to do a simple analytic solution benchmark of 2D geoid's density integral.

**Comment by ian-r-rose** [comment]
There are some analytical 2D geoid/topography benchmarks in the literature, so those will be good targets for testing. They are obviously significantly easier to run than 3D benchmarks (and we can even fake a delta function more easily, since we can go to much higher resolution).

Also, I suspect that the 3D global integral you are talking about is indeed very similar to the multipole approach. We can address this over  the next couple of days.

**Comment by Shangxin-Liu** [comment]
My 3D global integral is essentially the classical spherical harmonic integral to calculate density contribution but just transforms the equation a little bit to make it directly working in "no layer" 3D AMR. Now from the global refinement 1 coarse mesh test, it yields pretty nice accuracy compared to analytical solution. If Ian really wants to also implement another multipole approach, after benchmark maybe we can create an if loop to allow users to choose which method they want to use. @gassmoeller @ian-r-rose

**Pull request merged**